avro-converter/src/main/java/io/confluent/connect/avro/AvroData.java

```diff
@@ -933,13 +939,41 @@ private static Object maybeWrapSchemaless(Schema schema, Object value, String ty
         }
       }
 
+      // the new and correct way to handle logical types
+      if (schema.name() != null) {
+        if (Decimal.LOGICAL_NAME.equalsIgnoreCase(schema.name())) {
+          String precisionString = schema.parameters().get(CONNECT_AVRO_DECIMAL_PRECISION_PROP);
+          // preserve decimal logical type iff precision is specified,
+          // as assuming fixed precision for arbitrary precision record might incur data loss
+          if (precisionString != null) {
+            int precision = Integer.parseInt(precisionString);
+            int scale = Integer.parseInt(schema.parameters().get(Decimal.SCALE_FIELD));
+            org.apache.avro.LogicalTypes.decimal(precision, scale).addToSchema(baseSchema);
+          }
+        } else if (Time.LOGICAL_NAME.equalsIgnoreCase(schema.name())) {
+          org.apache.avro.LogicalTypes.timeMillis().addToSchema(baseSchema);
+        } else if (Timestamp.LOGICAL_NAME.equalsIgnoreCase(schema.name())) {
+          org.apache.avro.LogicalTypes.timestampMillis().addToSchema(baseSchema);
+        } else if (Date.LOGICAL_NAME.equalsIgnoreCase(schema.name())) {
+          org.apache.avro.LogicalTypes.date().addToSchema(baseSchema);
+        }
+      }
+
+      // Initially, to add support for logical types a new property was added
+      // with key `logicalType`. This enabled logical types for avro schemas but not others,
+      // such as parquet. The use of 'addToSchema` above supersedes this method here,
+      //  which should eventually be removed.
+      // Keeping for backwards compatibility until a major version upgrade happens.
+
+      // OLD COMMENT
       // Only Avro named types (record, enum, fixed) may contain namespace + name. Only Connect's
       // struct converts to one of those (record), so for everything else that has a name we store
       // the full name into a special property. For uniformity, we also duplicate this info into
       // the same field in records as well even though it will also be available in the namespace()
       // and name().
       if (schema.name() != null) {
-        if (Decimal.LOGICAL_NAME.equalsIgnoreCase(schema.name())) {
+        if (Decimal.LOGICAL_NAME.equalsIgnoreCase(schema.name())
```

Comment:
because now we requires that decimal logical type gets preserved iff the precision is specified. So, even for the old way, we might want to check if the precision is specified.

Reply:
I'd like us to leave the old logic unchanged in this PR

Reply:
On a second thought, let's keep them in sync with the above block